### PR TITLE
return an error when max retry is reached in executeMethod()

### PR DESCRIPTION
**Pull request by ericychoi**
DO NOT MERGE, missing tests.

### CODE REVIEW

**Comment by harshavardhana** [comment]
This is not a bad idea but i think we can return the actual error rather than sending special error like retry. Right now it is sent it is just that DeleteObject() ignores it. You can modify that part of code instead.

**Comment by ericychoi** [comment]
I believe `err` is not sent back to `RemoveObject()`.  Since `c.do()` will not return an error on 500.  (https://github.com/minio/minio-go/blob/master/api.go#L479)

It will return an error on closed connection or empty response, but not on a valid 500.  

Now I can inspect `resp` in `RemoveObject()` and if it contains non-2xx status code, return an error to signify that to the client.  Would you think that's better? Keep in mind then the same issue will keep happening for any other calls that use `executeMethod()` (update, for instance)

**Comment by harshavardhana** [comment]
> I believe err is not sent back to RemoveObject(). Since c.do() will not return an error on 500. (https://github.com/minio/minio-go/blob/master/api.go#L479)
> 
> It will return an error on closed connection or empty response, but not on a valid 500.
> 
> Now I can inspect resp in RemoveObject() and if it contains non-2xx status code, return an error to signify that to the client. Would you think that's better? Keep in mind then the same issue will keep happening for any other calls that use executeMethod() (update, for instance)

Yes in case of executeMethod() we should save the error regardless of the response codes.

File: api.go [hunk 1]
```diff
@@ -528,6 +532,11 @@ func (c Client) executeMethod(method string, metadata requestMetadata) (res *htt
 		// For all other cases break out of the retry loop.
 		break
 	}
+
+	if retryCount >= MaxRetry {
+		err = errors.New("Max retry reached")
```
Comment (dtjm) [inline api.go]: I think it's better to declare the error once at the package level, and return that instead.  

On the other hand, in this case, would you want to return the last error received?  I was thinking if you just return "Max retry reached", the caller won't have any idea what the underlying error was.

Comment (harshavardhana) [inline api.go]: Yes that is what we need here, we should only return the last error after all the retries.

**Comment by ericychoi** [comment]
Created #555 as an alternative approach.

**Comment by harshavardhana** [comment]
You can close this @ericychoi

**Comment by harshavardhana** [comment]
This approach is no longer needed.

**Pull request closed without merging**